// @slide-computer/signer/lib/cjs/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNetworkError = exports.isUserActionError = exports.isNotSupportedError = exports.isGeneralError = exports.NETWORK_ERROR = exports.ACTION_ABORTED_ERROR = exports.PERMISSION_NOT_GRANTED_ERROR = exports.NOT_SUPPORTED_ERROR = exports.GENERIC_ERROR = exports.isJsonRpcError = exports.isServerError = exports.INTERNAL_ERROR = exports.INVALID_PARAMS_ERROR = exports.METHOD_NOT_FOUND_ERROR = exports.INVALID_REQUEST_ERROR = exports.JSON_PARSE_ERROR = void 0;
// JSON RPC errors
exports.JSON_PARSE_ERROR = -32700;
exports.INVALID_REQUEST_ERROR = -32600;
exports.METHOD_NOT_FOUND_ERROR = -32601;
exports.INVALID_PARAMS_ERROR = -32602;
exports.INTERNAL_ERROR = -32603;
const isServerError = (error) => error >= -32099 && error <= -32000;
exports.isServerError = isServerError;
const isJsonRpcError = (error) => error >= -32768 && error <= -32000;
exports.isJsonRpcError = isJsonRpcError;
// ICRC-25 errors
exports.GENERIC_ERROR = 1000;
exports.NOT_SUPPORTED_ERROR = 2000;
exports.PERMISSION_NOT_GRANTED_ERROR = 3000;
exports.ACTION_ABORTED_ERROR = 3001;
exports.NETWORK_ERROR = 4000;
const isGeneralError = (error) => error >= 1000 && error <= 1999;
exports.isGeneralError = isGeneralError;
const isNotSupportedError = (error) => error >= 2000 && error <= 2999;
exports.isNotSupportedError = isNotSupportedError;
const isUserActionError = (error) => error >= 3000 && error <= 3999;
exports.isUserActionError = isUserActionError;
const isNetworkError = (error) => error >= 4000 && error <= 4999;
exports.isNetworkError = isNetworkError;
